File: packages/python/src/prefactor_evals_no_llm/result.py
"""The result object. Identical semantics in both languages."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Iterable, Optional

PASS = "pass"
FAIL = "fail"
SKIP = "skip"


@dataclass
class EvalResult:
    eval_id: str
    instance_id: str
    status: str
    details: str
    evidence: dict = field(default_factory=lambda: {"span_ids": [], "values": {}})
    # Only set on skip. Names the config key that would enable this eval, so the
    # scorecard can tell the user how to fix it rather than only that it broke.
    remedy: Optional[str] = None

    def to_dict(self) -> dict:
        out = {
            "eval_id": self.eval_id,
            "instance_id": self.instance_id,
            "status": self.status,
            "details": self.details,
            "evidence": self.evidence,
        }
        if self.remedy:
            out["remedy"] = self.remedy
        return out


def _evidence(span_ids: Optional[Iterable[str]], values: Optional[dict]) -> dict:
    return {"span_ids": list(span_ids or []), "values": dict(values or {})}


def passed(eval_id: str, instance_id: str, details: str,
           span_ids: Optional[Iterable[str]] = None,
           values: Optional[dict] = None) -> EvalResult:
    return EvalResult(eval_id, instance_id, PASS, details, _evidence(span_ids, values))


def failed(eval_id: str, instance_id: str, details: str,
           span_ids: Optional[Iterable[str]] = None,
           values: Optional[dict] = None) -> EvalResult:
    return EvalResult(eval_id, instance_id, FAIL, details, _evidence(span_ids, values))


def skipped(eval_id: str, instance_id: str, details: str,
            remedy: Optional[str] = None,
            values: Optional[dict] = None) -> EvalResult:
    """Skip is not a pass. It is counted separately everywhere.

    `remedy` should name the config key that would enable the eval. A scorecard
    that says "23 skipped" and stops there is useless.
    """
    return EvalResult(eval_id, instance_id, SKIP, details,
                      _evidence([], values), remedy=remedy)

File: packages/typescript/src/result.ts
/** The result object. Identical semantics and identical JSON shape to Python. */

export const PASS = 'pass';
export const FAIL = 'fail';
export const SKIP = 'skip';
export type Status = typeof PASS | typeof FAIL | typeof SKIP;

export interface Evidence {
  span_ids: string[];
  values: Record<string, unknown>;
}

/** Field names are snake_case to match the Python result JSON exactly. The
 * conformance suite diffs these objects, so a camelCase rename here would look
 * like a behaviour difference. */
export interface EvalResult {
  eval_id: string;
  instance_id: string;
  status: Status;
  details: string;
  evidence: Evidence;
  /** Only set on skip. Names the config key that would enable this check, so a
   * scorecard can say how to fix it rather than only that it did not run. */
  remedy?: string;
}

function evidence(spanIds?: readonly string[], values?: Record<string, unknown>): Evidence {
  return { span_ids: spanIds ? [...spanIds] : [], values: values ? { ...values } : {} };
}

export function passed(
  evalId: string, instanceId: string, details: string,
  spanIds?: readonly string[], values?: Record<string, unknown>,
): EvalResult {
  return {
    eval_id: evalId, instance_id: instanceId, status: PASS, details,
    evidence: evidence(spanIds, values),
  };
}

export function failed(
  evalId: string, instanceId: string, details: string,
  spanIds?: readonly string[], values?: Record<string, unknown>,
): EvalResult {
  return {
    eval_id: evalId, instance_id: instanceId, status: FAIL, details,
    evidence: evidence(spanIds, values),
  };
}

/** Skip is never a pass. It is counted separately everywhere. */
export function skipped(
  evalId: string, instanceId: string, details: string,
  remedy?: string, values?: Record<string, unknown>,
): EvalResult {
  return {
    eval_id: evalId, instance_id: instanceId, status: SKIP, details,
    evidence: evidence([], values), ...(remedy ? { remedy } : {}),
  };
}
